oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/SimpleFlatFileUtil.java

```diff
@@ -0,0 +1,109 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This util class can be used to export a tree (eg entire repository) to a flat
+ * file, without index dependency/involvement.
+ */
+public class SimpleFlatFileUtil {
+
+    private static final Logger log = LoggerFactory.getLogger(SimpleFlatFileUtil.class);
+
+    private static final String LINE_SEPARATOR = System.getProperty("line.separator");
+
+    private final FileWriter fw;
+    private final BufferedWriter bw;
+    private final ArrayList<StateInBytesHolder> entryBatch = new ArrayList<>();
+    private final Predicate<String> pathPredicate = path -> true;
+    private final NodeStateEntryWriter entryWriter;
+    private long totalLines = 0;
+
+    private SimpleFlatFileUtil(File f) throws IOException {
+        // blobStore is only used for deserialization - so pass null here:
+        entryWriter = new NodeStateEntryWriter(null);
+        fw = new FileWriter(f);
+        bw = new BufferedWriter(fw);
+    }
+
+    public static void createFlatFileFor(NodeState ns, File f) throws IOException {
+        final SimpleFlatFileUtil h = new SimpleFlatFileUtil(f);
+        log.info("createFlatFileFor : writing to {}", f.getCanonicalPath());
+        h.addEntryAndTraverseChildren(ns);
+        h.close();
+        log.info("createFlatFileFor : done. wrote {} lines in total.", h.totalLines);
+    }
+
+    private void close() throws IOException {
+        flush();
+        bw.close();
+        fw.close();
+    }
+
+    private void flush() throws IOException {
+        for (StateInBytesHolder nsh : entryBatch) {
+            String line = entryWriter.toString(nsh.getPathElements(), nsh.getLine());
+            bw.append(line);
+            bw.append(LINE_SEPARATOR);
+        }
+        log.info("flush : wrote another {} nodes, total so far: {} lines.",
+                entryBatch.size(), totalLines);
+        totalLines += entryBatch.size();
+        entryBatch.clear();
+    }
+
+    private void addEntryAndTraverseChildren(NodeState ns) throws IOException {
+        addEntry(ns);
+        if (entryBatch.size() > 999) {
+            flush();
+        }
+        for (ChildNodeEntry e : ns.getChildNodeEntries()) {
+            addEntryAndTraverseChildren(e.getNodeState());
+        }
+    }
+
+    private void addEntry(NodeState ns) {
+        DocumentNodeState dns = (DocumentNodeState) ns;
+        NodeStateEntry e = new NodeStateEntry.NodeStateEntryBuilder(dns,
+                dns.getPath().toString()).build();
+        String path = e.getPath();
+        if (!NodeStateUtils.isHiddenPath(path) && pathPredicate.test(path)) {
+            String jsonText = entryWriter.asJson(e.getNodeState());
+            // Here logic differs from NodeStateEntrySorter in sense that
+            // Holder line consist only of json and not 'path|json'
+            StateInBytesHolder h = new StateInBytesHolder(path, jsonText);
+            entryBatch.add(h);
```

Comment:
Why not writing the entry directly to the file instead of having an intermediate step of storing it in the `entryBatch` array? The code is using a buffered writer, so I don't see what is the advantage of having a second layer of buffering with this entryBatch array.

Reply:
+1, good point, I'll fix that.